### PR TITLE
Keep mobile keyboard on ENTER: delay focus until MCE init

### LINKED ISSUE #2601: Enter in iOS Safari hides keyboard
This breaks the writing flow. I have to re-focus the area to be able to type further.

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

This PR fixes #2601 and part of #4731. It delays moving the focus an editable that is still initialising, which fixes the keyboard hiding on ENTER.

## How Has This Been Tested?
Open Gutenberg in iOS with an on screen keyboard. Keyboard should not hide on ENTER.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by jasmussen** [comment]
Here's what I'm seeing in this branch: https://cloudup.com/cj4vQAOSyir

Are you seeing the same?

**Comment by iseulde** [comment]
@jasmussen So another issue seems to we the keyboard overlapping the target when it first pops up... I'll try to fix.

**Comment by iseulde** [comment]
It seems Gutenberg is in general suffering from a few scroll issues. When I open the inserter, the whole screen scrolls to the bottom.

**Comment by iseulde** [comment]
For our records, something similar happens with just plain contenteditables: https://codepen.io/iseulde/full/qxOoGd/

**Comment by iseulde** [comment]
Okay, I think this is a lot better now. We'll get the scroll container dynamically (changes based on viewport width at 600px). The scroll container must be padded so we can scroll into view when there is little content. Maybe there's a better solution here.

**Comment by jasmussen** [comment]
YOU DID IT! This experience is amazing. Not only did you fix the issue when creating new paragraphs, you fixed the bounce too, I didn't think that was possible. 

You deserve a _literal_ medal. These emoji don't do justice 🏅🥇🎖

**Comment by jasmussen** [comment]
Screen recordings. iOS looks pretty much perfect:

ios

Android seems to have a bit much padding at the bottom, it'd be nice if you could just see the previous paragraph above. 

android

I know it may not be polished yet, but how you solved this in two days boggles my mind.

**Comment by iseulde** [comment]
The only thing that bothers me still is that sometimes, when you focus some text that will later be covered by the keyboard, it won't scroll up.

**Comment by jasmussen** [comment]
> The only thing that bothers me still is that sometimes, when you focus some text that will later be covered by the keyboard, it won't scroll up.

On mobile Safari, the viewport doesn't change height when the soft keyboard shows up. It does on Android. I don't know how much there is we can do here.

**Comment by iseulde** [comment]
> The only thing that bothers me still is that sometimes, when you focus some text that will later be covered by the keyboard, it won't scroll up.

This issue is now fixed btw. :)

> Android seems to have a bit much padding at the bottom, it'd be nice if you could just see the previous paragraph above.

That's because the publish toolbar hides on iOS, but it does not on Android. Not sure how to fix ATM or what the mechanism that hides it on iOS is.

File: blocks/rich-text/index.js [hunk 1]
```diff
@@ -587,6 +618,13 @@ export default class RichText extends Component {
 				if ( event.shiftKey || ! this.props.onSplit ) {
 					this.editor.execCommand( 'InsertLineBreak', false, event );
 				} else {
+					// For small screens, save the root node rectangle so it can
+					// the position can be scrolled to in the next focussed
+					// instance.
+					if ( window.innerWidth < 784 ) {
+						scrollPosition = this.editor.getBody().getBoundingClientRect();
```
Comment (youknowriad) [inline blocks/rich-text/index.js]: This is probably a left over?

File: blocks/rich-text/index.js [hunk 2]
```diff
@@ -212,6 +219,30 @@ export default class RichText extends Component {
 	}
 
 	onFocus() {
+		// For small screens (virtual keyboard), always scroll the focussed
+		// editor into view. Unfortunately we cannot detect virtual keyboards.
+		if ( window.innerWidth < 784 ) {
```
Comment (youknowriad) [inline blocks/rich-text/index.js]: MAybe we could extract this to a module const to avoid repetition

File: blocks/rich-text/index.js [hunk 3]
```diff
@@ -884,6 +933,7 @@ export default class RichText extends Component {
 					{ ...ariaProps }
 					className={ className }
 					key={ key }
+					onMount={ this.onTinyMCEMount }
```
Comment (youknowriad) [inline blocks/rich-text/index.js]: Should we move these scrolling behaviors to the TinyMCE component instead of adding this Prop?

File: edit-post/components/layout/style.scss [hunk 4]
```diff
@@ -70,6 +70,7 @@
 
 .edit-post-layout__content {
 	position: relative;
+	padding-bottom: 100vh;
```
Comment (youknowriad) [inline edit-post/components/layout/style.scss]: What's the purpose of this change?

File: editor/components/block-list/block.js [hunk 5]
```diff
@@ -504,6 +480,7 @@ const mapStateToProps = ( state, { uid } ) => ( {
 	isFirstMultiSelected: isFirstMultiSelectedBlock( state, uid ),
 	isHovered: isBlockHovered( state, uid ) && ! isMultiSelecting( state ),
 	focus: getBlockFocus( state, uid ),
+	shouldFocusNext: shouldBlockFocusNext( state, uid ),
```
Comment (youknowriad) [inline editor/components/block-list/block.js]: So if I understand properly this new prop is useful because there's a new `dispatch` that will be performed later to move the focus? Why can't we batch these two dispatches together?

 - Maybe a unique actions
 - Using a store enhancer or a middleware (I believe these things already exist)

**Comment by youknowriad** [comment]
Added some reviewers as this kind of change is always tricky :)

File: blocks/rich-text/index.js (hunk 1 above)
Comment (iseulde) [reply blocks/rich-text/index.js]: What do you mean with left over? :)

File: blocks/rich-text/index.js (hunk 2 above)
Comment (iseulde) [reply blocks/rich-text/index.js]: Yeah, that's probably best. To utils?

File: blocks/rich-text/index.js (hunk 3 above)
Comment (iseulde) [reply blocks/rich-text/index.js]: I did, but then we need to pass the `scrollPosition` variable which is tied to this component. To me it looked cleaner this way, but I'm fine either way.

File: edit-post/components/layout/style.scss (hunk 4 above)
Comment (iseulde) [reply edit-post/components/layout/style.scss]: This could use a comment as well. :) To pad the scroll box if there is not enough content to scroll.

File: editor/components/block-list/block.js (hunk 5 above)
Comment (iseulde) [reply editor/components/block-list/block.js]: Yes, we let the editor that needs focus call `onFocus` by itself later. There's no newly added dispatch here, we are just moving focus too early. I'm not sure how an enhancer or middleware would help.

File: blocks/rich-text/index.js (hunk 1 above)
Comment (youknowriad) [reply blocks/rich-text/index.js]: Never mind, I though it was a local variable

File: blocks/rich-text/index.js (hunk 3 above)
Comment (youknowriad) [reply blocks/rich-text/index.js]: True, I missed the global variable

File: editor/store/reducer.js [hunk 6]
```diff
@@ -421,14 +422,13 @@ export function blockSelection( state = {
...
 			};
 		case 'INSERT_BLOCKS':
 			return {
 				...state,
-				start: action.blocks[ 0 ].uid,
-				end: action.blocks[ 0 ].uid,
-				focus: {},
-				isMultiSelecting: false,
+				// We must wait for Editable to initialise.
+				next: action.blocks[ 0 ].uid,
```
Comment (youknowriad) [inline editor/store/reducer.js]: Not sure I understand here, you're adding a `next` saying that this block will be focused next but for me `focus: {}` means the same thing?

So wondering why the `focus` prop is not passed at the same moment to the block without this change?

File: blocks/rich-text/index.js (hunk 2 above)
Comment (iseulde) [reply blocks/rich-text/index.js]: Hm, I just tried the iPad horizontally, and this won't cover it. We need a better way to detect virtual keyboards.

File: editor/store/reducer.js (hunk 6 above)
Comment (iseulde) [reply editor/store/reducer.js]: The `focus` props for the previous block will be taken away, which will cause the area to blur.

Comment (youknowriad) [reply editor/store/reducer.js]: So you want to focus before blur right? Maybe we can just delay the `.blur` call?

File: blocks/rich-text/index.js (hunk 3 above)
Comment (iseulde) [reply blocks/rich-text/index.js]: Global sounds bad :) I'd prefer to call it module variable ;)

File: editor/store/reducer.js (hunk 6 above)
Comment (iseulde) [reply editor/store/reducer.js]: How do we do that across instances? I guess we could maybe use something like `scrollPosition`. `scrollPosition || blur`.

Comment (youknowriad) [reply editor/store/reducer.js]: Actually, I wonder if this `.blur` called is needed at all, might be good to check the commit adding it.

Comment (youknowriad) [reply editor/store/reducer.js]: The `.blur()` call was added here https://github.com/WordPress/gutenberg/pull/683#issuecomment-299548842 

Where the reason some more "logical" than based on a use case (cc @aduth). If removing the blur fixed the mobile issue, I'd vote for that :)

Comment (iseulde) [reply editor/store/reducer.js]: Calling blur might be bad in any case, because the focus should always move somewhere else. This will just move it to the body element.

**Comment by iseulde** [comment]
I removed the blur call entirely, and while this seems to work on iPhone, the keyboard still hides on enter on iPad... Looking...

**Comment by iseulde** [comment]
This does not happen with the last version d1966db2f5de12135e7dd98eadc8abeb3343fac4... What is happing here... 😭

**Comment by iseulde** [comment]
@youknowriad It seems that it just sometimes works, sometimes it does not, because focus is first moved to the block wrapper:

https://github.com/WordPress/gutenberg/blob/56e192c4d0aefb733250b073f082823506135735/editor/components/block-list/block.js#L125-L128

With the `next` trick in the reducer, this focus call was skipped too. Somehow we need to remove this call if an editable is present?

**Comment by iseulde** [comment]
I'm also wondering why the autofocus is needed at all. I don't immediately see regressions taking it away. Having a look at the history here.

**Comment by youknowriad** [comment]
@iseulde I think this last call is related to arrow navigation across block (I'm pretty sure it's related to the writing flows)

**Comment by iseulde** [comment]
Yeah it actually makes sense focus moves there. But it shouldn't do that when there are editables :)

**Comment by iseulde** [comment]
screen shot 2018-02-02 at 17 33 55

This scenario doesn't work very well.

**Comment by iseulde** [comment]
I think the enter behaviour is beneficial for all screens, so I removed the check. It's one step toward the typewriter experience, and it will make sure it is tested in more scenarios. For the focus adjustment, it seems to be better to check the user agent instead of window width.

**Comment by jasmussen** [comment]
> I think the enter behaviour is beneficial for all screens, so I removed the check. It's one step toward the typewriter experience, and it will make sure it is tested in more scenarios.

I agree — and if it's as simple as a check to limit it to mobile again, we should absolutely push it to master and stress test it, so we can add learn.

**Comment by jasmussen** [comment]
> This scenario doesn't work very well.

That specific scenario is very hard to optimize for ;) — perhaps in the future when we get some more PWA features that could let us do some fullscreen stuff.

**Comment by iseulde** [comment]
> I agree — and if it's as simple as a check to limit it to mobile again, we should absolutely push it to master and stress test it, so we can add learn.

Yes, we can add the user agent check to it at any time.

There are some test error, but at first sight they seem unrelated to this PR.

**Comment by iseulde** [comment]
Rebased.

**Comment by iseulde** [comment]
Figuring out how to fix the e2e test.

**Comment by iseulde** [comment]
@youknowriad It seems the e2e tests just start typing too fast after click/enter. I always see the first character splitting off from the rest. Unsure how to fix at this time. I tried different selectors and delay.

**Comment by youknowriad** [comment]
ad part of #4872 I noticed that one of the tests was a false positive in master and the issue was surfaced by my changes (and I suspect it surfaced here as well). And it looks like there's no way to fix it at the moment so I commented and added a comment to explain the reasons.

Maybe if you rebase the tests will pass.

File: blocks/rich-text/index.js [hunk 7]
```diff
@@ -592,6 +616,11 @@ export default class RichText extends Component {
 				if ( event.shiftKey || ! this.props.onSplit ) {
 					this.editor.execCommand( 'InsertLineBreak', false, event );
 				} else {
+					// For type writing offect, save the root node offset so it
+					// can the position can be scrolled to in the next focussed
+					// instance.
```
Comment (mcsf) [inline blocks/rich-text/index.js]: Copy errors here.

File: blocks/rich-text/index.js [hunk 8]
```diff
@@ -212,6 +219,23 @@ export default class RichText extends Component {
 	}
 
 	onFocus() {
+		// For virtual keyboards, always scroll the focussed editor into view.
+		// Unfortunately we cannot detect virtual keyboards, so we check UA.
```
Comment (mcsf) [inline blocks/rich-text/index.js]: I don't have an external keyboard right now, but what happens if I use Gutenberg with a mobile device with an external keyboard? Most systems hide the virtual keyboard whenever an external keyboard is connected; any chance the auto-scrolling can be detrimental in those cases?

File: blocks/rich-text/index.js [hunk 9]
```diff
@@ -212,6 +219,23 @@ export default class RichText extends Component {
...
+		if ( /iPad|iPhone|iPod|Android/i.test( window.navigator.userAgent ) ) {
```
Comment (mcsf) [inline blocks/rich-text/index.js]: Is it worth making a `isMobile` util?

**Review by mcsf — COMMENTED** [review]
Nice work. :) Left some questions.

File: blocks/rich-text/index.js (hunk 8 above)
Comment (iseulde) [reply blocks/rich-text/index.js]: It wouldn't be _super_ annoying I think. The focussed field won't go out of view, and in these cases, we're still dealing with a relatively small screen. The bigger the screen, the worse this gets. While not ideal, this is the best I can come up with. I've played with other approaches as well, e.g. trying to detect touch (vs clicks), but failed to make it work. The order of which touch, click, and focus events are fired makes it seems impossible to do.

File: blocks/rich-text/index.js (hunk 9 above)
Comment (iseulde) [reply blocks/rich-text/index.js]: Maybe. Would be nice if there are more uses. In this specific case we're not really trying to detect "mobile" devices, but rather devices with virtual keyboards.

**Comment by iseulde** [comment]
I'm trying to rebase, but the last focus changes are causing some bugs here. Haven't figured out why yet, I may need to revisit this tomorrow.

File: blocks/rich-text/index.js (hunk 9 above)
Comment (mcsf) [reply blocks/rich-text/index.js]: Agreed. My thinking was also something like:

```js
isMobile = () => regex.test( … )
hasVirtualKeyboard = isMobile // best approximation
```

**Comment by mcsf** [comment]
> I'm trying to rebase, but the last focus changes are causing some bugs here. Haven't figured out why yet, I may need to revisit this tomorrow.

Let us know when/if you'd like a last look, once that's resolved.

**Comment by iseulde** [comment]
Looking at this again. Sometimes instead of scrolling down on enter, it scrolls up... I can't remember this before the rebase. Trying to debug but not sure what's going on.

**Comment by iseulde** [comment]
Got a bursting headache from this issue

**Comment by jorgefilipecosta** [comment]
There were some changes in parallel to this PR so conflicts happened. I rebased handled the conflicts and I did a new set of tests.
I think the work done here is awesome @iseulde! It's totally new experience writing on iOS after this PR. I did not found problems there and the problem of scrolling up on enter never happened (maybe the changes in parallel solved it).
The only regression I found is on the desktop, using arrows to scroll the editable paragraph may become out of view:
feb-27-2018 18-09-51
It looks like it is the only issue remaining I will try to find the reason for this behavior.

File: editor/components/block-list/block.js [hunk 10]
```diff
@@ -231,7 +231,10 @@ export class BlockListBlock extends Component {
 			return;
 		}
 
-		target.focus();
+		// If there is a contenteditable element, it will move focus by itself.
+		if ( ! target.querySelector( '[contenteditable="true"]' ) ) {
+			target.focus();
+		}
```
Comment (youknowriad) [inline editor/components/block-list/block.js]: This is the only change that's bothering me in this PR because it introduces a special case in how blocks focus is handled.

I'm thinking all this behavior `focusTabbable when a new block is selected` could be moved into a separate component and also handle the usecase we need for this PR: Focus the first tabbable element at the right position :)

---
Anyway, might not be that easy to do, so maybe for now, just consolidate the test of the presence of a TinyMCE editor to match the test already done here https://github.com/WordPress/gutenberg/pull/4775/files#diff-bc66d5701a2fea1111bb8f1bf0150910R244 `const editor = tinymce.get( target.getAttribute( 'id' ) );`

Comment (youknowriad) [reply editor/components/block-list/block.js]: Also, I believe this assumes that the TinyMCE will focus itself which I believe is not always what we want. Imagine inserting a block containing a plainText followed by the richText. TinyMCE won't be focused in that case.

I believe this is only necessary if we use the `onSplit` behavior of the RichText element.

File: blocks/rich-text/index.js [hunk 11]
```diff
@@ -541,6 +569,11 @@ export class RichText extends Component {
 				if ( event.shiftKey || ! this.props.onSplit ) {
 					this.editor.execCommand( 'InsertLineBreak', false, event );
 				} else {
+					// For type writing offect, save the root node offset so it
+					// can the position can be scrolled to in the next focussed
+					// instance.
+					offsetTop = rootNode.getBoundingClientRect().top;
```
Comment (youknowriad) [inline blocks/rich-text/index.js]: While this works at the moment, nothing guarantee that it will work in the future, imagine if the `onSplit` callback of the RichText component inserts something else that a `RichText`, say a `PlainText` or another unrelated block, how this would work? would it be a problem?

I don't have an alternative implementation at the moment but it seems fragile to rely on a global variable like this and assume a certain behavior in parent blocks.

**Comment by jasmussen** [comment]
Thank you for helping with this PR. It's going to improve the user experience so much, and it feels like it's nearly there, experience wise.

**Comment by notnownikki** [comment]
I found a similar (but slightly different) issue to https://github.com/WordPress/gutenberg/pull/4775#issuecomment-368972939

If I create a post and keep inserting paragraphs, then the space at the end of the post keeps growing and I can scroll the post out of view, whether I use the arrows or not.

outofview

This is in Chrome 61.

**Comment by notnownikki** [comment]
There's the same space at the bottom of posts on android too, you can swipe up and lose the post off the top of the screen.

**Comment by jasmussen** [comment]
Ella has done such a great job on this PR, and deserves multiple medals for so drastically improving the writing behavior on phones, notably the iPhone. She deserves multiple medals for this.

It would be a great thing if we were to pull together and ship this PR as soon as we can, as the way this fixes the phone experience would be great to test in master for a bit. It's also definitely clear that this is a technically difficult PR with many different moving pieces. How can we help ship this together, so it's not only on Ella's shoulders? 

Does it make sense to look at the individual pieces that make up this PR and ship them as smaller PRs separately? Is there a way we reduce this PR to a smaller, half-way step, and then pull together to ship the remaining pieces afterwards? CC: @aduth @youknowriad @gziolo @karmatosed @jorgefilipecosta

**Comment by iseulde** [comment]
Done in #5769.
